README.md
Project Overview:
-----------------

This project comes from the idea to propose an "out-of-the-box" system, as complete as possible, in both the management of different sensors and the implementation of different weather APIs. In this way, an ensemble of meteorological and local variables are centralised in a Raspberry PI (preferably a 4b with 4GB or more RAM), as well as different forecasts produced by public (ex: Météo-France) or private (ex: OpenWeather) entities.


More Information:
-----------------

The weather station is coded with Python 3.9 and relies on few technologies such as PostgreSQL, MQTT and Zigbee2MQTT.
It has been tested on a Raspberry PI 4b with Buster and Bullseye Raspberry PI OS.
Few APIs have been included : Météo-France and OpenWeatherMap. More will come in the future.
Few sensors are handled by the weather station : DS18B20 (one or more) and BME280 (one or more), through 1-wire and I2C buses. More will come in the future, once I can get my hand on them.
The main language of the software is French. English will be added once the full code is stable.


Features:
---------

* ephemeris of the day
* management of 1-Wire sensors (for now : DS18B20)
* management of I2C sensors (for now : BME280)
* management of MQTT-based sensors (temperature, humidity, pressure, battery and signal quality)
* management of weather APIs (for now : Météo-France, OpenWeatherMap)
* two different forecasts are handled : 1h forecasts and 6h/1day forecasts.
* forecasts are based on localisation (must be entered by the user)
* data are displayed as real time measurements and as times series on 24h.


Developments:
-------------

Develoments are made on my free time. As said above, the weather station software has been coded with Python 3.9, and thus it is pure Python. The GUI is handled by PyQt 5.15 and designed with QtCreator 8.0. The latest version of PyCharm is used to develop the software.


Installation:
-------------

The weather station system is a pure python script. Thus it can be used from a terminal, with the simple command : python3 weather_station.py.
At the present time, it is launched by *pi* user, but it can be used by any user as long as all rights have been set.

For developments or to run it as a script, it needs few dependancies.

Operating systems :

* Raspberry PI OS Buster

  or

* Raspberry PI OS Bullseye

Python modules :

* Python (3.7+)
* PyQt5 (5.15.0+)
* numpy (1.19.5+)
* matplotlib (3.3.0+)
* markdown (3.3.0+)
* requests (2.25.0+)
* psycopg2 (2.8.6+)
* pyowm (3.2.0+)
* meteofrance-api (1.0.2+)
* smbus2 (0.4.1+)
* RPi.bme280 (0.2.4+)
* paho.mqtt (1.6.1+)
* pigpio (1.78+) -> pigpiod service must me enabled
* pyephem (4.1.3+)
* pyqtspinner (0.1.1+)
* dirsync (2.2.5 +)

Other software :

* PostgreSQL (13.5+)
* Mosquitto (2.0.11+)
* Zigbee2Mqtt (1.24.0)

Fonts :

The weather station uses Source Sans Pro as a font. It needs to be installed to give better results.

Environment variable :

I use an environment variable ("WEATHER_STATION_DEV") on my development computers (Ubuntu and Windows). Thus it give the possibility to manipulate "fake" sensors with no access to smbus2/bme280/pigpio modules.

Stand-alone package:
--------------------

A stand-alone package is compiled using PyInstaller. Thus, a Python installation is not necessary. In that case, only PostgreSQL, Mosquitto and Zigbee2Mqtt are mandatory.

The weather station uses Source Sans Pro as a font. It needs to be installed to give better results.

Specific configuration for PostgreSQL, Mosquitto and Zigbee2Mqtt
----------------------------------------------------------------

For PosgreSQL, the following configuration must be used :

* a specific database must be created for the weather station with a dedicated user.
* read/write privileges on the database must be granted to the user.
* user, password, database and address must be entered in the Weather Station options.
* tables will be handled by the Weather Station.

For Mosquitto and Zigbee2Mqtt :

* the installation of Mosquitto and Zigbee2Mqtt have to be handled by the user. 
* the Weather Station uses a username, password, address and a maintopic (ex: zigbee2mqtt) to connect to mosquitto and subscribe to Zigbee2Mqtt topics. Thus the user must configure Mosquitto and Zigbee2Mqtt accordingly. The port in use is the default one.
* variables retrieve from Mosquitto through Zigbee2Mqtt must be set this way :
  * for each variable : MQTT/[maintopic]/[device]/[variable]
  * for exemple, temperature : MQTT/zigbee2mqtt/[device]/temperature
  * the [device] can be its serial number or its friendly name as set in Zigbee2Mqtt configuration file
* read the Zigbee2Mqtt documentation for a better explanation ;o) !


Sensors
-------

Currently, only BME280 and DS18B20 are officialy supported and tested. More will be supported once I can put my hand on them.
Sensors relying on MQTT to send and store data are supported if each variable has its own topic. At this time, only temperature, humidity, pressure, battery and signal quality are used by the weather station.


Limitations
-----------

The GUI has been developed on a raspberry pi 4 connected to a Waveshare 7" monitor (https://www.waveshare.com/product/raspberry-pi/displays/7inch-hdmi-lcd-c.htm?___SID=U) and is not responsive (lack of knowledge from the 
developer). Thus it can't adapt automatically to different screen size. It will be adapted in the next major version.

As the GUI is intended to be used with a touch screen, the mouse cursor is hidden on a linux system for all windows.

I found an issue between PyQt5, stylesheet and raspberry pi. Stylesheet and icons are not applied until those files are stored in the home directory of the user. That's why a sync function has been introduced to copy all icons and stylesheets in the home directory and keep those folders in sync. Once a proper solution is found, that limitation will be removed.


Documentation:
--------------

At this time, there is no documentation yet.


Screenshots:
------------

![Sensor parameters](screenshots/screenshot_1.png?raw=true "Parameters from different sensors") 

![Sensor time series](screenshots/screenshot_4.png?raw=true "Parameters can be displayed as time series")

![1h forecast](screenshots/screenshot_2.png?raw=true "Forecast every 1h on a 24h period")

![1d forecast](screenshots/screenshot_3.png?raw=true "Forecast every day on a 5d period")

(displayed parameters are randomly generated for development purpose)


documentation/changelog.txt

### July 20 2023, Release version 0.16.0 ###

  * MODIFIED:
    * MQTT manager window has been modified to handle new MQTT main topic.

  * ADDED:
    * a new function to update the sensor dictionary has been added.
    * all functions to handle new MQTT main topic has been added.

  * FIXED:
    * fixed missing modules


### July 20 2023, Release version 0.15.2 ###

  * MODIFIED:
    * log message, when closing the gui, has been modified to register values of reboot and shutdown variables.
    * new version of pyqtspinner.
    * code has been cleaned.


### October 10 2022, Release version 0.15.1 ###

  * FIXED:
    * an issue was found and fixed in the thread to compute ephemeris.
    * the reboot function after an update wasn't working properly.


### October 10 2022, Release version 0.15.0 ###

  * FIXED:
    * the computation of the moon phase has been corrected.

  * MODIFIED:
    * all qthreads with a run section has been changed to include a qtimer and allow proper termination.


### September 09 2022, Release version 0.14.0 ###

  * MODIFIED:
    * removed old code in utils.py.
    * if no connexion or no forecast can be parsed, a no data icon is now displayed in forecast sections, and labels are
     cleared.
     * the function to update the config file has been changed for better performance.
     * the creation and update of the config file have been updated with the new API.
     * the plot function has been moved to a new thread, and the spinner has been re-added.
     * a new separator icon has been introduced.

  * ADDED:
    * a function to update a configuration file has been introduced.
    * few new options have been added to the configuration file.
    * new options have been added to the option window.
    * it is now possible to ask the weather station to test automatically the connectivity every period of time
    defined by the user if it is offline.
    * new options for in and out temperature/humidity/pressure, have been added to the option window.
    * it is now possible to choose which sensor will feed temperature/humidity/pressure data.
    * the sensor which feeds temperature data will feed battery and signal data.
    * an error message has been added if all options concerning database are empty.

  * FIXED:
    * a widget has been forgotten in keyboard window.
    * if no sensors and remote sensors are registered, the sensor threads are none and not well cleared when the gui is
    closed.
    * json import has been deleted.


### August 20 2022, Release version 0.13.0 ###

  * FIXED:
    * when exporting data, if an error occurred and a retry is done, the text in the warning box is not cleared.
    * two variables weren't initialized correctly.

  * ADDED :
    * gusts have been added to 1h forecast window.
    * gusts have been added to 6h forecast window.
    * gusts have been added to 1d OpenWeather forecast window.
    * an icon was missing.
    * an icon for gusts has been added.
    * an icon to open file has been added.
    * if sensor data are old (>2h), a message is displayed.
    * a new infowindow has been introduced.
    * a new way to obtain elevation has been introduced.

  * MODIFIED:
    * the stylesheet of few windows have been modified for better consistency with the main window.
    * the 1d forecast page with MF forecast now show data based on daily forecast.
    * a new way to plot time series has been introduced to improve rendering times.
    * font files have been removed. Now Source Sans Pro must be installed on Raspberry Pi.
    * units for different fields have been put in their own QLabel widget
    * improved reading of battery, signal, pressure, temperature and humidity windows.


### June 23 2022, Release version 0.12.2 ###

  * FIXED:
    * plot thread tried to compute min and max of a list even if the list is empty.
    * MQTT test thread was not handled properly since last version.
    * the dot icons were not well refreshed when the time series button is clicked.
    * MQTT threads was crashing when fields other than temperature were not available for MQTT devices.
    * the mqtt device manager tried to parse calibration methode and value even it is not handled yet. It has been fixed
    .

  * ADDED :
    * a button and a function to export database data have been introduced.


### May 5 2022, Release version 0.12.1 ###

  * MODIFIED:
    * forecast data are now automatically parse and displayed once they are received from weather API.

functions/utils.py
import collections
import json
import shutil
import logging
import pathlib
import datetime
import platform
import configparser
from dirsync import sync
from logging.handlers import RotatingFileHandler
from PyQt5 import QtGui, QtWidgets


def create_option_file(user_path):
    config_dict = configparser.ConfigParser()
    config_dict['LOG'] = {'level': 'DEBUG', 'path': str(user_path)}
    config_dict['SYSTEM'] = {'place_altitude': '', 'check_update': 'False', 'auto_check_connexion': 'False',
                             'auto_connexion_unit': 'minutes', 'auto_connexion_value': ''}
    config_dict['API'] = {'api_used': 'meteofrance', 'api_key': '', 'user_place': 'False', 'request_rate': '30'}
    config_dict['DISPLAY'] = {'in_display_rate': '30', 'in_temperature': '', 'in_humidity': '', 'in_pressure': '',
                              'in_msl_pressure': 'False', 'out_display_rate': '30', 'out_temperature': '',
                              'out_humidity': '', 'out_pressure': '', 'out_msl_pressure': 'False', 'in_bat_signal': '',
                              'out_bat_signal': ''}
    config_dict['SENSOR'] = {'sensors_rate': '30'}
    config_dict['TIMESERIES'] = {'in_temperature': '', 'in_humidity': '', 'out_temperature': '', 'out_pressure': '',
                                 'msl_pressure': 'False'}
    config_dict['DATABASE'] = {'port': '', 'username': '', 'password': '', 'database': '', 'host': ''}
    ini_file = open(pathlib.Path(user_path).joinpath('weather_station.ini'), 'w')
    config_dict.write(ini_file)
    ini_file.close()


def update_config_file(user_path):
    ini_path = pathlib.Path(user_path).joinpath('weather_station.ini')
    option_missing, remove_option = False, False
    config_dict = configparser.ConfigParser()
    config_dict.read(ini_path)
    new_option_list = {'SYSTEM': {'auto_check_connexion': 'False',
                                  'auto_connexion_unit': 'minutes',
                                  'auto_connexion_value': ''},
                       'DISPLAY': {'in_temperature': '', 'in_humidity': '', 'in_pressure': '', 'in_bat_signal': '',
                                   'out_temperature': '', 'out_humidity': '', 'out_pressure': '', 'out_bat_signal': ''}}
    old_options_list = {'DISPLAY': ['in_sensor', 'out_sensor']}
    for section, options in new_option_list.items():
        if section not in list(config_dict.keys()):
            option_missing = True
            config_dict[section] = {}
        for option, value in options.items():
            if option not in list(config_dict[section].keys()):
                config_dict[section][option] = value
    for section, options in old_options_list.items():
        for option in options:
            config_dict.remove_option(section, option)
            remove_option = True
            if not config_dict.items(section):
                config_dict.remove_section(section)
    if option_missing or remove_option:
        ini_file = open(ini_path, 'w')
        config_dict.write(ini_file)
        ini_file.close()


def create_sensor_file(user_path):
    date = datetime.datetime.now().strftime('%d/%m/%Y')
    json_dict = {'creation_date': date, 'modification_date': date, 'DS18B20': {}, 'BME280': {},
                 'MQTT': {'username': '', 'password': '', 'address': '', 'port': '', 'main_topics': [], 'devices': {}}}
    f = open(pathlib.Path(user_path).joinpath('sensor_file.json'), 'w')
    json.dump(json_dict, f, indent=4)
    f.close()


def update_sensor_file(user_path):
    f = open(pathlib.Path(user_path).joinpath('sensor_file.json'), 'r')
    json_dict = json.load(f)
    f.close()
    option_missing = False

    if 'port' not in list(json_dict['MQTT'].keys()):
        json_dict['MQTT']['port'] = '1883'
        option_missing = True

    if 'main_topics' not in list(json_dict['MQTT'].keys()):
        json_dict['MQTT']['main_topics'] = ['']
        option_missing = True

    if option_missing:
        f = open(pathlib.Path(user_path).joinpath('sensor_file.json'), 'w')
        json.dump(json_dict, f, indent=4)
        f.close()


def create_logging_handlers(config_dict, filename, default_path):
    using_default_path = False
    if pathlib.Path(config_dict.get('LOG', 'path')).exists():
        log_filename = str(pathlib.Path(config_dict.get('LOG', 'path')).joinpath(filename))
    else:
        using_default_path = True
        log_filename = str(pathlib.Path(default_path).joinpath(filename))
    logging.getLogger('').handlers = []
    logging.basicConfig(handlers=[RotatingFileHandler(log_filename, maxBytes=2000000, backupCount=5)],
                        level=getattr(logging, config_dict.get('LOG', 'level')),
                        format='%(asctime)s : %(levelname)s : %(message)s')
    logging.getLogger('matplotlib').setLevel(logging.WARNING)
    formatter = logging.Formatter('%(levelname)s : %(message)s')
    console = logging.StreamHandler()
    console.setLevel(logging.DEBUG)
    console.setFormatter(formatter)
    logging.getLogger('').addHandler(console)
    if using_default_path:
        logging.error('gui - logging system - path from ini file not found, using default path')


def sync_graphic_folders(gui_path):
    logging.debug(f'gui - utils.py - sync_graphic_folders - gui_path: {gui_path}')
    if platform.system() == 'Linux':
        path = pathlib.Path().home()
        gui_path = pathlib.Path(gui_path)
        if path.joinpath('icons').exists():
            logging.debug('gui - utils.py - sync_graphic_folders - synchronizing graphic folders')
            sync(str(gui_path.joinpath('icons')), path.joinpath('icons'), 'sync')
            sync(str(gui_path.joinpath('graphic_materials')), path.joinpath('graphic_materials'), 'sync')
        else:
            logging.debug('gui - utils.py - sync_graphic_folders - coping graphic folders')
            shutil.copytree(str(gui_path.joinpath('icons')), path.joinpath('icons'))
            shutil.copytree(str(gui_path.joinpath('graphic_materials')), path.joinpath('graphic_materials'))


def clear_layout(layout):
    logging.debug('gui - utils.py - clear_layout')
    for i in reversed(range(layout.count())):
        item = layout.itemAt(i)
        if isinstance(item, QtWidgets.QWidgetItem):
            item.widget().deleteLater()
        elif isinstance(item, QtWidgets.QLayout):
            clear_layout(item.layout())
        layout.removeItem(item)


def icon_creation_function(icon_filename):
    logging.debug(f'gui - utils.py - icon_creation_function - icon_filename: {icon_filename}')
    icon = QtGui.QIcon()
    icon.addPixmap(QtGui.QPixmap(f'icons/{icon_filename}'), QtGui.QIcon.Normal, QtGui.QIcon.Off)
    icon.addPixmap(QtGui.QPixmap(f'icons/{icon_filename}'), QtGui.QIcon.Disabled, QtGui.QIcon.Off)
    icon.addPixmap(QtGui.QPixmap(f'icons/{icon_filename}'), QtGui.QIcon.Disabled, QtGui.QIcon.On)
    return icon


def pictogramme_creation_function(icon_filename):
    logging.debug(f'gui - utils.py - icon_creation_function - icon_filename: {icon_filename}')
    path = f'graphic_materials/pictogrammes/{icon_filename}'
    icon = QtGui.QIcon()
    icon.addPixmap(QtGui.QPixmap(path), QtGui.QIcon.Normal, QtGui.QIcon.Off)
    return icon


def font_creation_function(font_style):
    logging.debug(f'gui - utils.py - font_creation_function - font_style: {font_style}')
    font = QtGui.QFont()
    font.setKerning(True)
    font.setStyleStrategy(QtGui.QFont.PreferAntialias)
    font.setFamily("Source Sans Pro")
    if font_style == 'normal':
        font.setPointSize(11)
    elif font_style == 'big':
        font.setFamily("Source Sans Pro Semibold")
        font.setPointSize(11)
    elif font_style == 'small':
        font.setPointSize(10)
    elif font_style == 'small-italic':
        font.setPointSize(10)
        font.setItalic(True)
    elif font_style == 'very_big':
        font.setPointSize(18)
    elif font_style == 'medium_big':
        font.setPointSize(14)
    return font


def stylesheet_creation_function(stylesheet):
    logging.debug(f'gui - utils.py - stylesheet_creation_function - stylesheet: {stylesheet}')
    f = open(f'graphic_materials/style_sheets/{stylesheet}_stylesheet.dat', 'r')
    stylesheet_str = ''.join(f.readlines())
    f.close()
    return stylesheet_str


def shadow_creation_function(offset, blur):
    logging.debug(f'gui - utils.py - shadow_creation_function - offset: {offset} ; blur: {blur}')
    shadow = QtWidgets.QGraphicsDropShadowEffect()
    shadow.setOffset(offset)
    shadow.setBlurRadius(blur)
    return shadow


def days_months_dictionary():
    logging.debug('gui - utils.py - days_months_dictionary')
    date_dict = {'day': {1: 'Lundi', 2: 'Mardi', 3: 'Mercredi', 4: 'Jeudi', 5: 'Vendredi', 6: 'Samedi', 7: 'Dimanche'},
                 'month': {1: 'Janvier', 2: 'Février', 3: 'Mars', 4: 'Avril', 5: 'Mai', 6: 'Juin', 7: 'Juillet',
                           8: 'Août', 9: 'Septembre', 10: 'Octobre', 11: 'Novembre', 12: 'Décembre'}}
    return date_dict


def weather_to_pictogrammes(weather, dt=None, sunrise=None, sunset=None):
    logging.debug(f'gui - utils.py - weather_to_pictogrammes - weather: {weather} ; sunrise: {sunrise} ; sunset: '
                  f'{sunset}')
    available_weather = {'Eclaircies': 'eclaircies.svg',
                         'Très nuageux': 'couvert.svg',
                         'Ensoleillé': 'ensoleille.svg',
                         'Risque d\'orages': 'risque_orages.svg',
                         'Nuit claire': 'nuit_claire.svg',
                         'Couvert': 'couvert.svg',
                         'Rares averses': 'rares_averses.svg',
                         'Pluies éparses': 'pluies_eparses.svg',
                         'Risque de grêle': 'risque_grele.svg',
                         'Averses orageuses': 'averses_orageuses.svg',
                         'Brume': 'brume.svg',
                         'Averses': 'pluie.svg',
                         'Pluie': 'pluie.svg',
                         'Pluies orageuses': 'averses_orageuses.svg',
                         'Orages': 'orages.svg',
                         'Ciel voilé': 'ciel_voile.svg',
                         'Bancs de Brouillard': 'brume.svg',
                         'Brouillard': 'brouillard.svg',
                         'Brouillard givrant': 'brouillard_givrant.svg',
                         'Neige': 'neige.svg',
                         'Averses de neige': 'neige.svg',
                         'Pluie et neige': 'pluie_et_neige.svg',
                         'Quelques flocons': 'quelques_flocons.svg'}
    try:
        pictogramme = available_weather[weather]
    except KeyError:
        logging.warning('gui - utils.py - weather_to_pictogrammes - pictogramme is missing : ' + weather)
        pictogramme = ''

    if sunrise is None or sunset is None or dt is None:
        period = 'jour'
    else:
        if isinstance(sunset, list):
            if sunset[0] <= dt <= sunrise[1]:
                period = 'nuit'
            else:
                period = 'jour'
        else:
            if sunrise <= dt <= sunset:
                period = 'jour'
            else:
                period = 'nuit'
    if pictogramme:
        link = f'graphic_materials/pictogrammes/{period}/{pictogramme}'
    else:
        link = 'icons/none_icon.png'
    icon = QtGui.QIcon()
    icon.addPixmap(QtGui.QPixmap(link), QtGui.QIcon.Normal, QtGui.QIcon.Off)
    return icon


def openweather_to_mf_desc(item):
    logging.debug(f'gui - utils.py - openweather_to_mf_desc - item : {item}')
    conversion = {'800n': 'Nuit claire', '800d': 'Ensoleillé', '801d': 'Eclaircies', '801n': 'Eclaircies',
                  '802d': 'Très nuageux', '802n': 'Très nuageux', '803d': 'Très nuageux',
                  '803n': 'Très nuageux', '804d': 'Couvert', '804n': 'Couvert', '500d': 'Pluies éparses',
                  '500n': 'Pluies éparses', '501d': 'Pluie', '501n': 'Pluie', '502d': 'Pluie', '502n': 'Pluie',
                  '503d': 'Pluie', '503n': 'Pluie', '504d': 'Pluie', '504n': 'Pluie',
                  '511d': 'Pluies verglaçantes', '511n': 'Pluies verglaçantes',
                  '520d': 'Rares averses', '520n': 'Rares averses', '521d': 'Averses', '521n': 'Averses',
                  '522d': 'Averses', '522n': 'Averses', '531d': 'Averses', '531n': 'Averses',
                  '701d': 'Brume', '701n': 'Brume', '711d': 'Brouillard', '711n': 'Brouillard',
                  '721d': 'Brouillard', '721n': 'Brouillard', '731d': 'Brouillard', '731n': 'Brouillard',
                  '741d': 'Brouillard', '741n': 'Brouillard', '751d': 'Brouillard', '751n': 'Brouillard',
                  '761d': 'Brouillard', '761n': 'Brouillard', '762d': 'Brouillard', '762n': 'Brouillard',
                  '771d': 'Brouillard', '771n': 'Brouillard', '781d': 'Brouillard', '781n': 'Brouillard',
                  '200d': 'Pluies orageuses', '200n': 'Pluies orageuses', '201d': 'Averses orageuses',
                  '201n': 'Averses orageuses', '202d': 'Averses orageuses', '202n': 'Averses orageuses',
                  '210d': 'Risque d\'orages', '210n': 'Risque d\'orages', '211d': 'Orages', '211n': 'Orages',
                  '212d': 'Orages', '212n': 'Orages', '221d': 'Orages', '221n': 'Orages',
                  '230d': 'Risque de grèle', '230n': 'Risque de grèle', '231d': 'Risque de grèle',
                  '231n': 'Risque de grèle', '232d': 'Risque de grèle', '232n': 'Risque de grèle',
                  '300d': 'Averses', '300n': 'Averses', '301d': 'Averses', '301n': 'Averses',
                  '302d': 'Averses', '302n': 'Averses', '310d': 'Averses', '310n': 'Averses',
                  '311d': 'Averses', '311n': 'Averses', '312d': 'Averses', '312n': 'Averses',
                  '313d': 'Averses', '313n': 'Averses', '314d': 'Averses', '314n': 'Averses',
                  '321d': 'Averses', '321n': 'Averses', '600d': 'Neige', '601d': 'Neige', '602d': 'Neige',
                  '611d': 'Pluie et neige', '612d': 'Pluie et neige', '613d': 'Pluie et neige',
                  '615d': 'Pluie et neige', '616d': 'Pluie et neige', '621d': 'Neige', '622d': 'Neige',
                  '623d': 'Neige', '600n': 'Neige', '601n': 'Neige', '602n': 'Neige', '611n': 'Pluie et neige',
                  '612n': 'Pluie et neige',  '613n': 'Pluie et neige', '615n': 'Pluie et neige',
                  '616n': 'Pluie et neige', '621n': 'Neige', '622n': 'Neige', '623n': 'Neige'}

    try:
        weather = conversion[item]
    except KeyError:
        weather = 'none'
    return weather


def wind_dir_to_pictogramme(dir_idx):
    logging.debug(f'gui - utils.py - wind_dir_to_pictogramme - dir_idx: {dir_idx}')
    path = f'graphic_materials/pictogrammes/wind_dir_images/wind_dir_arrow_{dir_idx}.svg'
    icon = QtGui.QIcon()
    icon.addPixmap(QtGui.QPixmap(path), QtGui.QIcon.Normal, QtGui.QIcon.Off)
    return icon


def code_to_departement():
    logging.debug('gui - utils.py - code_to_departement')
    code_dict = {
        '01': 'Ain',
        '02': 'Aisne',
        '03': 'Allier',
        '04': 'Alpes-de-Haute-Provence',
        '05': 'Hautes-Alpes',
        '06': 'Alpes-Maritimes',
        '07': 'Ardèche',
        '08': 'Ardennes',
        '09': 'Ariège',
        '10': 'Aube',
        '11': 'Aude',
        '12': 'Aveyron',
        '13': 'Bouches-du-Rhône',
        '14': 'Calvados',
        '15': 'Cantal',
        '16': 'Charente',
        '17': 'Charente-Maritime',
        '18': 'Cher',
        '19': 'Corrèze',
        '2A': 'Corse-du-Sud',
        '2B': 'Haute-Corse',
        '21': 'Côte-d\'Or',
        '22': 'Côtes-d\'Armor',
        '23': 'Creuse',
        '24': 'Dordogne',
        '25': 'Doubs',
        '26': 'Drôme',
        '27': 'Eure',
        '28': 'Eure-et-Loir',
        '29': 'Finistère',
        '30': 'Gard',
        '31': 'Haute-Garonne',
        '32': 'Gers',
        '33': 'Gironde',
        '34': 'Hérault',
        '35': 'Ille-et-Vilaine',
        '36': 'Indre',
        '37': 'Indre-et-Loire',
        '38': 'Isère',
        '39': 'Jura',
        '40': 'Landes',
        '41': 'Loir-et-Cher',
        '42': 'Loire',
        '43': 'Haute-Loire',
        '44': 'Loire-Atlantique',
        '45': 'Loiret',
        '46': 'Lot',
        '47': 'Lot-et-Garonne',
        '48': 'Lozère',
        '49': 'Maine-et-Loire',
        '50': 'Manche',
        '51': 'Marne',
        '52': 'Haute-Marne',
        '53': 'Mayenne',
        '54': 'Meurthe-et-Moselle',
        '55': 'Meuse',
        '56': 'Morbihan',
        '57': 'Moselle',
        '58': 'Nièvre',
        '59': 'Nord',
        '60': 'Oise',
        '61': 'Orne',
        '62': 'Pas-de-Calais',
        '63': 'Puy-de-Dôme',
        '64': 'Pyrénées-Atlantiques',
        '65': 'Hautes-Pyrénées',
        '66': 'Pyrénées-Orientales',
        '67': 'Bas-Rhin',
        '68': 'Haut-Rhin',
        '69': 'Rhône',
        '70': 'Haute-Saône',
        '71': 'Saône-et-Loire',
        '72': 'Sarthe',
        '73': 'Savoie',
        '74': 'Haute-Savoie',
        '75': 'Paris',
        '76': 'Seine-Maritime',
        '77': 'Seine-et-Marne',
        '78': 'Yvelines',
        '79': 'Deux-Sèvres',
        '80': 'Somme',
        '81': 'Tarn',
        '82': 'Tarn-et-Garonne',
        '83': 'Var',
        '84': 'Vaucluse',
        '85': 'Vendée',
        '86': 'Vienne',
        '87': 'Haute-Vienne',
        '88': 'Vosges',
        '89': 'Yonne',
        '90': 'Territoire de Belfort',
        '91': 'Essonne',
        '92': 'Hauts-de-Seine',
        '93': 'Seine-Saint-Denis',
        '94': 'Val-de-Marne',
        '95': 'Val-d\'Oise',
        '971': 'Guadeloupe',
        '972': 'Martinique',
        '973': 'Guyane',
        '974': 'La Réunion',
        '976': 'Mayotte'}
    return code_dict


def mpl_hour_list():
    logging.debug('gui - utils.py - mpl_hour_list')
    now = datetime.datetime.now()
    hours_list = [now]
    for h in range(1, 25, 1):
        new_hour = now - datetime.timedelta(hours=h)
        hours_list.append(new_hour)
    return list(reversed(hours_list))


def define_time_ticks(now):
    tick_list, label_list = [], []
    i = 0
    for hour in reversed(list(range(1, 25, 1))):
        tick_list.append(now - datetime.timedelta(hours=hour))
        if i == 0:
            label_list.append(f'-{hour}h')
        else:
            label_list.append('')
        i += 1
        if i == 4:
            i = 0
    tick_list.append(now)
    label_list.append('Now')
    return tick_list, label_list


def set_size(bytes_size):
    logging.debug('gui - utils.py - set_size')
    suffixes = ['B', 'KB', 'MB', 'GB', 'TB', 'PB']
    i = 0
    while bytes_size >= 1024 and i < len(suffixes) - 1:
        bytes_size /= 1024.
        i += 1
    f = ('%.2f' % bytes_size).rstrip('0').rstrip('.')
    idx = f.find('.')
    if idx >= 0:
        if len(f[idx:]) == 2:
            f += '0'
        elif len(f[idx:]) == 1:
            f += '00'
    else:
        f += '.00'
    return f'{f} {suffixes[i]}'


def battery_value_icon_dict():
    logging.debug('gui - utils.py - battery_value_icon_dict')
    data_dict = {0: 'batterie_0-5_icon.svg', 10: 'batterie_1-5_icon.svg', 30: 'batterie_2-5_icon.svg',
                 50: 'batterie_3-5_icon.svg', 70: 'batterie_4-5_icon.svg', 90: 'batterie_full_icon.svg',
                 101: 'batterie_full_icon.svg'}
    return data_dict


def link_value_icon_dict():
    logging.debug('gui - utils.py - link_value_icon_dict')
    data_dict = {0: 'signal_0-5_icon.svg', 40: 'signal_1-5_icon.svg', 80: 'signal_2-5_icon.svg',
                 120: 'signal_3-5_icon.svg', 160: 'signal_4-5_icon.svg', 200: 'signal_full_icon.svg',
                 256: 'signal_full_icon.svg'}
    return data_dict


def angle_moon_phase():
    logging.debug('gui - utils.py - angle_moon_phase')
    angle_dict = collections.OrderedDict()
    angle_dict[0.] = 'wi-moon-alt-new.svg'
    angle_dict[6.428571] = 'wi-moon-alt-waning-crescent-6.svg'
    angle_dict[19.285713] = 'wi-moon-alt-waning-crescent-5.svg'
    angle_dict[32.142855] = 'wi-moon-alt-waning-crescent-4.svg'
    angle_dict[44.999997] = 'wi-moon-alt-waning-crescent-3.svg'
    angle_dict[57.857138] = 'wi-moon-alt-waning-crescent-2.svg'
    angle_dict[70.714281] = 'wi-moon-alt-waning-crescent-1.svg'
    angle_dict[83.571423] = 'wi-moon-alt-third-quarter.svg'
    angle_dict[96.428565] = 'wi-moon-alt-waning-gibbous-6.svg'
    angle_dict[109.285707] = 'wi-moon-alt-waning-gibbous-5.svg'
    angle_dict[122.142849] = 'wi-moon-alt-waning-gibbous-4.svg'
    angle_dict[134.999991] = 'wi-moon-alt-waning-gibbous-3.svg'
    angle_dict[147.857133] = 'wi-moon-alt-waning-gibbous-2.svg'
    angle_dict[160.714275] = 'wi-moon-alt-waning-gibbous-1.svg'
    angle_dict[173.571417] = 'wi-moon-alt-full.svg'
    angle_dict[186.428559] = 'wi-moon-alt-waxing-gibbous-6.svg'
    angle_dict[199.285701] = 'wi-moon-alt-waxing-gibbous-5.svg'
    angle_dict[212.142843] = 'wi-moon-alt-waxing-gibbous-4.svg'
    angle_dict[224.999985] = 'wi-moon-alt-waxing-gibbous-3.svg'
    angle_dict[237.857127] = 'wi-moon-alt-waxing-gibbous-2.svg'
    angle_dict[250.714269] = 'wi-moon-alt-waxing-gibbous-1.svg'
    angle_dict[263.571411] = 'wi-moon-alt-first-quarter.svg'
    angle_dict[276.428553] = 'wi-moon-alt-waxing-crescent-6.svg'
    angle_dict[289.285695] = 'wi-moon-alt-waxing-crescent-5.svg'
    angle_dict[302.142837] = 'wi-moon-alt-waxing-crescent-4.svg'
    angle_dict[314.999979] = 'wi-moon-alt-waxing-crescent-3.svg'
    angle_dict[327.857121] = 'wi-moon-alt-waxing-crescent-2.svg'
    angle_dict[340.714263] = 'wi-moon-alt-waxing-crescent-1.svg'
    angle_dict[353.571405] = 'wi-moon-alt-new.svg'
    angle_dict[360.] = None
    return angle_dict


def get_season(dt):
    logging.debug(f'gui - utils.py - get_season - dt: {dt}')
    if isinstance(dt, datetime.datetime):
        dt = dt.date()
    y = dt.year
    seasons = [('Hiver', (datetime.date(y, 1, 1), datetime.date(y, 3, 20))),
               ('Printemps', (datetime.date(y, 3, 21), datetime.date(y, 6, 20))),
               ('Eté', (datetime.date(y, 6, 21), datetime.date(y, 9, 22))),
               ('Automne', (datetime.date(y, 9, 23), datetime.date(y, 12, 20))),
               ('Hiver', (datetime.date(y, 12, 21), datetime.date(y, 12, 31)))]
    return next(season for season, (start, end) in seasons if start <= dt <= end)


def str2bool(string):
    logging.debug(f'gui - utils.py - str2bool - string: {string}')
    if string == 'True':
        return True
    elif string == 'False':
        return False
    else:
        raise ValueError
